File: src/Kremeing.Core/Geo.fs
namespace Kremeing.Core

open System
open Kremeing.Contracts.Domain

/// Great-circle distance helpers. The `/stores/nearby` proxy gets a
/// `distance` straight from upstream, but the device-push fan-out needs
/// to compute "is this flipped store within the subscriber's radius?"
/// locally, against coordinates we already hold — hence a self-contained
/// haversine with no external dependency.
module Geo =

    /// Mean Earth radius in miles. Good to a few tenths of a percent for
    /// the "within N miles" proximity checks we use it for.
    [<Literal>]
    let EarthRadiusMiles = 3958.7613

    let private toRadians (degrees: float) = degrees * Math.PI / 180.0

    /// Haversine great-circle distance between two coordinates, in miles.
    /// Symmetric and always non-negative; returns 0 for identical points.
    let distanceMiles (a: Coordinates) (b: Coordinates) : float =
        let dLat = toRadians (b.Latitude - a.Latitude)
        let dLng = toRadians (b.Longitude - a.Longitude)
        let lat1 = toRadians a.Latitude
        let lat2 = toRadians b.Latitude
        let h =
            Math.Sin(dLat / 2.0) ** 2.0
            + Math.Cos lat1 * Math.Cos lat2 * Math.Sin(dLng / 2.0) ** 2.0
        // Clamp to guard against tiny FP overshoot past 1.0 feeding asin.
        2.0 * EarthRadiusMiles * Math.Asin(min 1.0 (sqrt h))

    /// True when `point` is within (inclusive) `radiusMiles` of `center`.
    let withinRadius (center: Coordinates) (radiusMiles: float) (point: Coordinates) : bool =
        distanceMiles center point <= radiusMiles

File: android/logic/src/main/kotlin/com/kremeing/auto/logic/Geo.kt
package com.kremeing.auto.logic

import kotlin.math.asin
import kotlin.math.cos
import kotlin.math.sin
import kotlin.math.sqrt

/**
 * Great-circle distance helpers. Kept in parity with the backend's
 * `Geo` module (same Earth radius in miles, same haversine formula) so the
 * app can rank/filter by distance identically to the server when it only has
 * raw coordinates (e.g. from an FCM payload) rather than a precomputed
 * `distanceMiles`.
 */
object Geo {
    /** Earth's mean radius in miles — must match the backend constant. */
    const val EARTH_RADIUS_MILES: Double = 3958.7613

    /** Haversine distance in miles between two lat/lng points. */
    fun distanceMiles(
        lat1: Double,
        lng1: Double,
        lat2: Double,
        lng2: Double,
    ): Double {
        val dLat = Math.toRadians(lat2 - lat1)
        val dLng = Math.toRadians(lng2 - lng1)
        val rLat1 = Math.toRadians(lat1)
        val rLat2 = Math.toRadians(lat2)
        val a = sin(dLat / 2) * sin(dLat / 2) +
            cos(rLat1) * cos(rLat2) * sin(dLng / 2) * sin(dLng / 2)
        return 2 * EARTH_RADIUS_MILES * asin(sqrt(a))
    }

    /** Inclusive radius test, mirroring the backend `Geo.withinRadius`. */
    fun withinRadius(
        lat1: Double,
        lng1: Double,
        lat2: Double,
        lng2: Double,
        radiusMiles: Double,
    ): Boolean = distanceMiles(lat1, lng1, lat2, lng2) <= radiusMiles
}
